### PR TITLE
SPARKC-387: Have columntype dependent on C* Protocol Version

### PR DESCRIPTION
**Pull request by RussellSpitzer**
There is an error where the connector will attempt to use types that are not valid for the C* version which is connected to. This patch makes all Type decisions protocol version aware to fix backwards compatibility.

### CODE REVIEW

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/types/ColumnType.scala [hunk 1]
```diff
@@ -157,7 +153,13 @@ object ColumnType {
   }
 
   /** Returns natural Cassandra type for representing data of the given Scala type */
-  def fromScalaType(dataType: Type): ColumnType[_] = {
+  def fromScalaType(
+    dataType: Type,
+    protocolVersion: ProtocolVersion = ProtocolVersion.NEWEST_SUPPORTED): ColumnType[_] = {
+
+    import ProtocolVersion._
+    def versionGTE( targetProtocolVersion: ProtocolVersion) : Boolean =
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/types/ColumnType.scala]: not very scala-ish... :)
maybe a value class with > and  < methods would be better?

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/RefCountedCache.scala [hunk 2]
```diff
@@ -28,7 +30,7 @@ final class RefCountedCache[K, V](create: K => V,
   private[cql] val deferredReleases = new TrieMap[V, ReleaseTask]
 
   private def createNewValueAndKeys(key: K): (V, Set[K]) = {
-    val value = create(key)
+    val value = RetryUtil.TryUntilValid(create(key))
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/RefCountedCache.scala]: why ?

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/RDDFunctions.scala [hunk 3]
```diff
@@ -92,7 +92,9 @@ class RDDFunctions[T](rdd: RDD[T]) extends WritableToCassandra[T] with Serializa
     rwf: RowWriterFactory[T],
     columnMapper: ColumnMapper[T]): Unit = {
 
-    val table = TableDef.fromType[T](keyspaceName, tableName)
+    val protocolVersion = connector.withClusterDo(_.getConfiguration.getProtocolOptions.getProtocolVersion)
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/RDDFunctions.scala]: so why it is not retrieved from the configuration ? why do we have that configuration entry?

Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/RDDFunctions.scala]: This is the "actual" protocol version being used. The Configuration entry is for forcing an alternate PV. The Default "null" uses the "HighestSupportedVersion" by the driver.

Connecting to C* 3.X with the 3.0 Driver and the Parameter == null yields PV4
Connecting to C* 3.X with the 3.0 Driver and the Parameter == PV3 yields PV3
Connecting to C* 2.1.X with the 3.0 Driver and the Parameter == null yields PV3

I can remove this now though as I decided to just gate the whole test suite

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/types/ColumnType.scala (hunk 1 above)
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/types/ColumnType.scala]: It implements comparable? I could have that wrapped I think?

**Comment by RussellSpitzer** [comment]
@jacek-lewandowski I think this is ready for review now, in the 2.1.15 build we should have 2 Failing tests, which I can fix in https://datastax-oss.atlassian.net/projects/SPARKC/issues/SPARKC-402 As these are not regressions. They are simply bugs in the code that previously existed

**Comment by jtgrabowski** [comment]
+1

**Comment by jtgrabowski** [comment]
can we merge this together with SPARKC-402 so that we have repo history without commits that do not build?

**Comment by RussellSpitzer** [comment]
Merging with SPARKC-402

**Pull request closed without merging**